### PR TITLE
Speaker embs

### PR DESCRIPTION
**Pull request by nithinraok**
Adds

- extraction of speaker embeddings to logging version folder
- Voxceleb evaluation file

Fixes:
- Restore_from .nemo when class expects arguments that aren't passed through cfg file

### CODE REVIEW

File: nemo/collections/asr/models/label_models.py [hunk 1]
```diff
@@ -165,8 +180,48 @@ def validation_epoch_end(self, outputs):
...
-        train_acc = (sum([x['n_correct_pred'] for x in outputs]) / sum(x['n_pred'] for x in outputs)) * 100
-        logging.info("training accuracy {:.3f}".format(train_acc))
+    def test_step(self, batch, batch_ix):
+        self.eval()
+        audio_signal, audio_signal_len, labels, _ = batch
+        _, embs = self.forward(input_signal=audio_signal, input_signal_length=audio_signal_len)
+        return {'embs': embs, 'labels': labels}
+
+
+class ExtractSpeakerEmbeddingsModel(EncDecSpeakerLabelModel):
```
Comment (okuchaiev) [inline nemo/collections/asr/models/label_models.py:190]: could you please add docstring here?

File: nemo/collections/asr/models/label_models.py [hunk 2]
```diff
@@ -165,8 +180,48 @@ def validation_epoch_end(self, outputs):
...
+    def __init__(self, cfg: DictConfig, trainer: Trainer = None, root_dir='.'):
```
Comment (okuchaiev) [inline nemo/collections/asr/models/label_models.py:191]: model's constructor should only take cfg and (optionally) trainer. Please remove root_dir from constructor args
Can't you just take root_dir from cfg?

File: nemo/core/classes/common.py [hunk 3]
```diff
@@ -247,8 +247,7 @@ def from_config_dict(cls, config: DictConfig, *args: Any, **kwargs: Any):
             instance = hydra.utils.instantiate(config=config, *args, **kwargs)
         else:
             # models are handled differently for now
-            # TODO: allow passthrough for args, and kwargs too?
-            instance = cls(cfg=config)
+            instance = cls(cfg=config, *args, **kwargs)
```
Comment (okuchaiev) [inline nemo/core/classes/common.py:250]: please revert this change for now

File: nemo/core/classes/modelPT.py [hunk 4]
```diff
@@ -166,7 +166,7 @@ def save_to(self, save_path: str):
             self.__make_nemo_file_from_folder(filename=save_path, source_dir=tmpdir)
 
     @classmethod
-    def restore_from(cls, restore_path: str):
+    def restore_from(cls, restore_path: str, *args: Any, **kwargs: Any):
```
Comment (okuchaiev) [inline nemo/core/classes/modelPT.py:169]: why? could you please change this back to the original signature

File: nemo/core/classes/modelPT.py [hunk 5]
```diff
@@ -190,8 +190,8 @@ def restore_from(cls, restore_path: str):
             model_weights = path.join(tmpdir, _MODEL_WEIGHTS)
             conf = OmegaConf.load(config_yaml)
             OmegaConf.set_struct(conf, True)
-            instance = cls.from_config_dict(config=conf)
-            instance.load_state_dict(torch.load(model_weights))
+            instance = cls.from_config_dict(config=conf, *args, **kwargs)
```
Comment (okuchaiev) [inline nemo/core/classes/modelPT.py:193]: same here

File: nemo/collections/asr/models/label_models.py (hunk 2 above)
Comment (nithinraok) [reply nemo/collections/asr/models/label_models.py:191]: I could, but anyhow considering the version formatting from PTL and hydra, I would have to anyhow overwrite them, so passing this as an arg.

File: nemo/core/classes/common.py (hunk 3 above)
Comment (nithinraok) [reply nemo/core/classes/common.py:250]: This change is helping me restoring model else I am getting garbage values

File: nemo/core/classes/modelPT.py (hunk 4 above)
Comment (nithinraok) [reply nemo/core/classes/modelPT.py:169]: To facilitate for optional root_dir , else this is not necessary

File: examples/speaker_recognition/conf/config.yaml [hunk 6]
```diff
@@ -7,17 +7,25 @@ n_filters: &n_filters 512
 
 model:
   train_ds:
-    manifest_filepath: '/data/manifests/voxceleb1/train/train_manifest.json'
+    manifest_filepath: '/data/an4_speaker/train.json'
```
Comment (titu1994) [inline examples/speaker_recognition/conf/config.yaml:10]: Use ??? here and pass path explicitly.

File: examples/speaker_recognition/conf/config.yaml [hunk 7]
```diff
@@ -7,17 +7,25 @@ n_filters: &n_filters 512
...
     sample_rate: 16000
     labels: null
     batch_size: 64 
     shuffle: True
     time_length: 8
 
   validation_ds:
-    manifest_filepath: '/data/manifests/voxceleb1/train/dev_manifest.json'
+    manifest_filepath: '/data/an4_speaker/dev.json'
```
Comment (titu1994) [inline examples/speaker_recognition/conf/config.yaml:18]: Same as above

File: examples/speaker_recognition/conf/config.yaml [hunk 8]
```diff
@@ -7,17 +7,25 @@ n_filters: &n_filters 512
...
     sample_rate: 16000
-    labels: None
+    labels: null
+    batch_size: 128
+    shuffle: False
+    time_length: 8
+
+  test_ds:
+    manifest_filepath: '/data/an4_speaker/test.json'
```
Comment (titu1994) [inline examples/speaker_recognition/conf/config.yaml:26]: Same as above

File: examples/speaker_recognition/speaker_reco.py [hunk 9]
```diff
@@ -43,15 +44,20 @@
 
 """
 
+seed_everything(42)
```
Comment (titu1994) [inline examples/speaker_recognition/speaker_reco.py:47]: @okuchaiev Do we want to encourage seeding? I believe some kernels of cuDNN anyway will be non-deterministic if you use cudnn autotune kernels.

File: examples/speaker_recognition/speaker_reco.py [hunk 10]
```diff
@@ -43,15 +44,20 @@
...
 
     logging.info(f'Hydra config: {cfg.pretty()}')
     trainer = pl.Trainer(**cfg.trainer)
-    exp_manager(trainer, cfg.get("exp_manager", None))
+    log_dir = exp_manager(trainer, cfg.get("exp_manager", None))
     speaker_model = EncDecSpeakerLabelModel(cfg=cfg.model, trainer=trainer)
     trainer.fit(speaker_model)
+    model_path = os.path.join(log_dir, '..', 'spkr.nemo')
+    speaker_model.save_to(model_path)
+    trainer.test(speaker_model)
```
Comment (titu1994) [inline examples/speaker_recognition/speaker_reco.py:60]: May want to run test on a new Trainer with a single GPU to get correct scores + no effects of DDP. Look at how speech recognition example handles this.

File: examples/speaker_recognition/spkr_get_emb.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+import pytorch_lightning as pl
+from pytorch_lightning import seed_everything
+
+from nemo.collections.asr.models import ExtractSpeakerEmbeddingsModel
+from nemo.core.config import hydra_runner
+from nemo.utils import logging
+from nemo.utils.exp_manager import exp_manager
+
+os.environ["TOKENIZERS_PARALLELISM"] = "false"
```
Comment (titu1994) [inline examples/speaker_recognition/spkr_get_emb.py:25]: Dont go setting this in code, it overrides user override from command line. Pass it explicitly in the execution example.

File: examples/speaker_recognition/spkr_get_emb.py [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+        hydra.run.dir="." \
+        trainer.gpus=0 \
+        trainer.max_epochs=50
+
+
+Add PyTorch Lightning Trainer arguments from CLI:
+    python speaker_reco.py \
+        ... \
+        +trainer.fast_dev_run=true
+
```
Comment (titu1994) [inline examples/speaker_recognition/spkr_get_emb.py:41]: What is this above example used for? Just to try it out ?

File: examples/speaker_recognition/spkr_get_emb.py [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+Hydra logs will be found in "$(./outputs/$(date +"%y-%m-%d")/$(date +"%H-%M-%S")/.hydra)"
+PTL logs will be found in "$(./outputs/$(date +"%y-%m-%d")/$(date +"%H-%M-%S")/lightning_logs)"
+
+"""
+
+seed_everything(42)
```
Comment (titu1994) [inline examples/speaker_recognition/spkr_get_emb.py:47]: I dont know if we want to seed.

File: examples/speaker_recognition/spkr_get_emb.py [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+@hydra_runner(config_path="conf", config_name="config")
+def main(cfg):
+
+    logging.info(f'Hydra config: {cfg.pretty()}')
+    trainer = pl.Trainer(**cfg.trainer)
+    log_dir = exp_manager(trainer, cfg.get("exp_manager", None))
+    model_path = os.path.join(log_dir, '..', 'spkr.nemo')
+    speaker_model = ExtractSpeakerEmbeddingsModel.restore_from(model_path, root_dir=log_dir)
+    trainer.test(speaker_model)
```
Comment (titu1994) [inline examples/speaker_recognition/spkr_get_emb.py:58]: Follow testing procedure from speech recognition code and run it with default Trainer

File: examples/speaker_recognition/spkr_get_emb.py [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+
+    # TODO fix this restoring issue later once sufficient fix is provided
+    # speaker_model = ExtractSpeakerEmbeddingsModel(cfg=cfg.model, trainer=trainer, root_dir= log_dir)
+    # speaker_model.restore_from(model_path)
```
Comment (titu1994) [inline examples/speaker_recognition/spkr_get_emb.py:62]: Remove this for now (and mark it for later correction). User facing code should not have todos.

File: examples/speaker_recognition/voxceleb_eval.py [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+    )
+
+    args = parser.parse_args()
+    trial_file, emb, manifest = args.trial_file, args.emb, args.manifest
+
+    y_score, y = get_acc(trial_file=trial_file, emb=emb, manifest=manifest)
+    fpr, tpr, thresholds = roc_curve(y, y_score, pos_label=1)
+
+    eer = brentq(lambda x: 1.0 - x - interp1d(fpr, tpr)(x), 0.0, 1.0)
+    # print("EER: {:.2f}%".format(eer * 100))
```
Comment (titu1994) [inline examples/speaker_recognition/voxceleb_eval.py:126]: Remove print maybe? Below syswrite does the same.

File: nemo/collections/asr/models/label_models.py [hunk 17]
```diff
@@ -147,6 +156,12 @@ def training_step(self, batch, batch_nb):
 
         return {'loss': loss_value, 'log': tensorboard_logs, "n_correct_pred": n_correct_pred, "n_pred": len(labels)}
 
+    def training_epoch_end(self, outputs):
```
Comment (titu1994) [inline nemo/collections/asr/models/label_models.py:159]: Why log to stdout but not to wandb/tensorboard?

File: nemo/collections/asr/models/label_models.py (hunk 2 above)
Comment (titu1994) [reply nemo/collections/asr/models/label_models.py:191]: @nithinraok Dont have a different signature than the rest of the codebase. Do whatever override of the config you need internally, but signature of constructor can accept only cfg and trainer.

File: nemo/collections/asr/models/label_models.py [hunk 18]
```diff
@@ -165,8 +180,48 @@ def validation_epoch_end(self, outputs):
...
+        return {'embs': embs, 'labels': labels}
+
+
+class ExtractSpeakerEmbeddingsModel(EncDecSpeakerLabelModel):
+    def __init__(self, cfg: DictConfig, trainer: Trainer = None, root_dir='.'):
+        super().__init__(cfg=cfg, trainer=trainer)
+        self.dir = root_dir
+
+    def test_step(self, batch, batch_ix):
+        self.eval()
```
Comment (titu1994) [inline nemo/collections/asr/models/label_models.py:196]: Not needed, PTL does this for you.

File: nemo/collections/asr/models/label_models.py [hunk 19]
```diff
@@ -165,8 +180,48 @@ def validation_epoch_end(self, outputs):
...
+        audio_signal, audio_signal_len, labels, _ = batch
+        _, embs = self.forward(input_signal=audio_signal, input_signal_length=audio_signal_len)
+        return {'embs': embs, 'labels': labels}
+
+    def test_epoch_end(self, outputs):
+        embs = torch.cat([x['embs'] for x in outputs])
+        # print(embs)
```
Comment (titu1994) [inline nemo/collections/asr/models/label_models.py:203]: Remove print

File: nemo/core/classes/common.py (hunk 3 above)
Comment (titu1994) [reply nemo/core/classes/common.py:250]: Can you inspect what are the values of *args and **kwargs. What special values are going into your `cls` that with them it gets proper embeddings but without it results in garbage values.

File: nemo/core/classes/modelPT.py (hunk 4 above)
Comment (titu1994) [reply nemo/core/classes/modelPT.py:169]: @nithinraok Dont pass root dir like this. Manipulate your own config and your code if needed but dont edit signature of core components which is shared across multiple models. Also, *args and **kwargs will not be allowed in any case as it breaks graphs in some form.

File: examples/speaker_recognition/conf/config.yaml (hunk 6 above)
Comment (nithinraok) [reply examples/speaker_recognition/conf/config.yaml:10]: that is what I was thinking but having some root_dir passed to model in cfg, is something i'm not comfortable with.

Comment (titu1994) [reply examples/speaker_recognition/conf/config.yaml:10]: I don't understand the issue maybe, The rest of the models follow the same approach.

File: examples/speaker_recognition/voxceleb_eval.py [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+        Y = X_test[y_idx]
+
+        if x_speaker not in keys:
+            keys.append(x_speaker)
+            trial_embs.extend([X])
+
+        if y_speaker not in keys:
+            keys.append(y_speaker)
+            trial_embs.extend([Y])
+
```
Comment (jbalam-nv) [inline examples/speaker_recognition/voxceleb_eval.py:92]: Looks like keys and trial_embs are only used for KALDI PLDA. can you pass an argument like save_kaldi_emb and put this part of the code under `if save_kaldi_emb:` . You could also uncomment the code below and put it under the same condition. 

May be also store keys and trial_embs in a dict, instead of traversing the keys list every line?

File: examples/speaker_recognition/voxceleb_eval.py [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+Args:
+    trial_file str: path to voxceleb trial file
+    emb : path to numpy file of embeddings (generated from spkr_get_emb.py)
+    manifest: path to test manifest file that contains path of audio files which is used in spkr_get_emb.py
+    Note: order of audio files in manifest file should match the embeddings
+"""
+
+
+def get_labels(manifest):
+    lines = open(manifest, 'r').readlines()
```
Comment (jbalam-nv) [inline examples/speaker_recognition/voxceleb_eval.py:39]: minor nit: you could use `with open(...)` to open files or make sure you close them. Should not really matter much in python, but still :)

File: examples/speaker_recognition/voxceleb_eval.py [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+        if uniq_name in test_list:
+            raise KeyError("uniq name is already present")
+        test_list[uniq_name] = idx
+    return test_list
+
+
+def get_acc(trial_file='', emb='', manifest=''):
+
+    X_test = np.load(emb)
+    manifest_lines = open(manifest, 'r').readlines()
```
Comment (jbalam-nv) [inline examples/speaker_recognition/voxceleb_eval.py:55]: `with open()` ?

File: examples/speaker_recognition/voxceleb_eval.py [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+
+    tmp_file = open(trial_file, 'r').readlines()
+    trail_score = open('trial_score.txt', 'w')
+
+    trial_embs = []
+    keys = []
+    all_scores = []
+    all_keys = []
+
+    # for each of trails in trial file
```
Comment (jbalam-nv) [inline examples/speaker_recognition/voxceleb_eval.py:68]: trails->trials

File: examples/speaker_recognition/speaker_reco.py (hunk 10 above)
Comment (nithinraok) [reply examples/speaker_recognition/speaker_reco.py:60]: why there are effects of DDP on test but not on valid?

File: examples/speaker_recognition/spkr_get_emb.py (hunk 14 above)
Comment (nithinraok) [reply examples/speaker_recognition/spkr_get_emb.py:58]: If I pass default trainer, my trainer object would't have exp_manager info, which will crash exp_manager call, @blisc

Comment (titu1994) [reply examples/speaker_recognition/spkr_get_emb.py:58]: Have you looked at the speech recognition example ? It also uses exp manager and doesn't crash.

Comment (titu1994) [reply examples/speaker_recognition/spkr_get_emb.py:58]: Can you put a stack trace of the crash ?

File: examples/speaker_recognition/voxceleb_eval.py (hunk 22 above)
Comment (nithinraok) [reply examples/speaker_recognition/voxceleb_eval.py]: Done

File: examples/speaker_recognition/voxceleb_eval.py (hunk 23 above)
Comment (nithinraok) [reply examples/speaker_recognition/voxceleb_eval.py]: Done

File: examples/speaker_recognition/voxceleb_eval.py (hunk 21 above)
Comment (nithinraok) [reply examples/speaker_recognition/voxceleb_eval.py]: Done

File: examples/speaker_recognition/voxceleb_eval.py (hunk 20 above)
Comment (nithinraok) [reply examples/speaker_recognition/voxceleb_eval.py:72]: Thanks for this suggestion, was planning to move embs to dict and key based, done now. Added save_kaldi_emb flag too

File: examples/speaker_recognition/spkr_get_emb.py (hunk 14 above)
Comment (nithinraok) [reply examples/speaker_recognition/spkr_get_emb.py]: Resolved

File: examples/speaker_recognition/voxceleb_eval.py (hunk 20 above)
Comment (jbalam-nv) [reply examples/speaker_recognition/voxceleb_eval.py:72]: I think you could also use trial_emb_dict and use it like this:
`if save_kaldi_emb and x_speaker not in trial_emb_dict:
                trial_emb_dict[x_speaker] = X`
instead of list lookup `x_speaker not in keys` .

File: examples/speaker_recognition/speaker_reco.py (hunk 9 above)
Comment (titu1994) [reply examples/speaker_recognition/speaker_reco.py:45]: I suppose autotune is turned off by default so no harm done

File: examples/speaker_recognition/spkr_get_emb.py (hunk 11 above)
Comment (titu1994) [reply examples/speaker_recognition/spkr_get_emb.py:25]: At the very lease check here if the user has already set this value, and default to false if they have not.

File: nemo/collections/asr/models/label_models.py (hunk 17 above)
Comment (titu1994) [reply nemo/collections/asr/models/label_models.py:153]: Do we want double logging via logging.info as well as wandb and tensorboard ? This code runs on more than one process, so printed logs are bound to be out of order. For train logs, just maybe use the tensorboard/wandb logs.

**Comment by nithinraok** [comment]
> Overall looks good now. Minor comments about overriding command line level `TOKENIZERS_PARALLELISM` and logging in train step.

 tested  from command line but forgot to remove this. Thanks
